Add custom setuptools command for 'develop' (v3)

ana opened the pull request:
Changes from v3 of https://github.com/avocado-framework/avocado/pull/4725

* Add the option --external to develop that will install and  uninstall external plugins.
* extra ball: Use the list-based version of run() for the manpage command, so there is no need for a shell.

beraldoleal commented on setup.py:
```diff
@@ -86,6 +86,16 @@ def clean_optional_plugins():
 class Develop(setuptools.command.develop.develop):
     """Custom develop command."""
 
+    user_options = setuptools.command.develop.develop.user_options + [
```
Have you tried this?

```python
self.user_options + [....
```

ana replied:
That doesn't work, 'self' is not defined there.

beraldoleal replied:
Sure! I could swear that this was inside a method. ;)

beraldoleal commented on setup.py:
```diff
@@ -86,6 +86,16 @@ def clean_optional_plugins():
+        ("external", None, "Install external plugins in development mode"),
+    ]
+
+    boolean_options = setuptools.command.develop.develop.boolean_options + ['external']
```
same here.

beraldoleal commented on setup.py:
```diff
@@ -178,7 +178,7 @@ def run(self):
 
         try:
-            run('{} man/avocado.rst man/avocado.1'.format(cmd), shell=True, check=True)
+            run([cmd, "man/avocado.rst", "man/avocado.1"], check=True)
```
Nice!

beraldoleal reviewed (commented):
Hi @ana, this LGTM, just one small nitpick comment.

Thank you for accepting the suggestions.

beraldoleal commented:
We also have one error with python 3.6.

ana commented:
> We also have one error with python 3.6.

It build fine after triggering a rebuild. Looks like GHA also has its hiccups.

ana commented:
I found why it's failing in cirrus. The command `python setup.py develop --user` before PR was only installing avocado.
With this PR  the command is installing the plugins and the setup.py of every plugin is installing its dependencies. The very first in the list `html` needs `jinja2` and the containers fail to download it (because the certificate).

Once https://github.com/avocado-framework/avocado/pull/4732 is merged, I expect the problem to be solved for Debian at least.

beraldoleal reviewed (approved):
LGTM, thanks!

clebergnu reviewed (changes requested):
Hi @ana, this looks great, but it has one unintended (or at least different) behavior when compared to the previous feature.  If you look at the documentation here:

https://avocado-framework.readthedocs.io/en/89.0/guides/contributor/chapters/environment.html#installing-in-develop-mode

It describes the intended use.  The most important bits are:

```
$ cd $AVOCADO_PROJECTS_DIR
$ git clone $AVOCADO_GIT
$ git clone $AVOCADO_PROJECT2
```

That means that both Avocado, and other projects/plugins/extensions, are put at `$AVOCADO_PROJECTS_DIR`.  Previously, only a *single* level from `$AVOCADO_PROJECTS_DIR` was searched for `setup.py` files.  Now, it seems to search for `setup.py` everywhere.

The unintended end result is that example plugins in avocado `examples/plugins/*`, and anything else, is actually enabled.  This is probably an easy fix, but better to keep the current behavior IMO.

ana commented:
> Hi @ana, this looks great, but it has one unintended (or at least different) behavior when compared to the previous feature. If you look at the documentation here:
> 
> https://avocado-framework.readthedocs.io/en/89.0/guides/contributor/chapters/environment.html#installing-in-develop-mode
> 
> It describes the intended use. The most important bits are:
> 
> ```
> $ cd $AVOCADO_PROJECTS_DIR
> $ git clone $AVOCADO_GIT
> $ git clone $AVOCADO_PROJECT2
> ```
> 
> That means that both Avocado, and other projects/plugins/extensions, are put at `$AVOCADO_PROJECTS_DIR`. Previously, only a _single_ level from `$AVOCADO_PROJECTS_DIR` was searched for `setup.py` files. Now, it seems to search for `setup.py` everywhere.
> 
> The unintended end result is that example plugins in avocado `examples/plugins/*`, and anything else, is actually enabled. This is probably an easy fix, but better to keep the current behavior IMO.

Hi @clebergnu 

Yes, the current behavior is slightly different to the one I'm proposing as you point out. And I can change that. However I think we need to update the documentation because the current behavior is also problematic, if I clone both avocado and avocado-vt in the home directory and export this directory as `AVOCADO_EXTERNAL_PLUGINS_PATH`, when I  run `make external-develop` I get:

```
....
>> LINK /home/ana/avocado-vt
>> SKIP /home/ana/.config
>> SKIP /home/ana/.vim
>> LINK /home/ana/avocado
>> FAIL /home/ana/avocado
>> SKIP /home/ana/.cache
>> SKIP /home/ana/.local
>> SKIP /home/ana/.pylint.d
```

Thoughts?

ana commented:
> However I think we need to update the documentation because the current behavior is also problematic,
> ....
> Thoughts?

I have done some reading about the plugins split and I think the good solution  here is tied to that. We could change the documentation to tell the user to export `AVOCADO_EXTERNAL_PLUGINS_PATH` to the directory containing the git checkout of all the plugins.

ana commented:
@clebergnu  I changed `rglob` by `glob`, that should keep the current behavior in the Makefile

clebergnu commented:
> @clebergnu I changed `rglob` by `glob`, that should keep the current behavior in the Makefile

I've resubmitted a few failing CI jobs (failed when communicating with PyPI).

ana commented:
> > @clebergnu I changed `rglob` by `glob`, that should keep the current behavior in the Makefile
> 
> I've resubmitted a few failing CI jobs (failed when communicating with PyPI).

You have an explanation to that problem in a previous comment https://github.com/avocado-framework/avocado/pull/4731#issuecomment-868683661
I have added a new commit with a fix  for debian and ubuntu, I don't know how to fix in windows.

clebergnu commented:
> > > @clebergnu I changed `rglob` by `glob`, that should keep the current behavior in the Makefile
> > 
> > 
> > I've resubmitted a few failing CI jobs (failed when communicating with PyPI).
> 
> You have an explanation to that problem in a previous comment [#4731 (comment)](https://github.com/avocado-framework/avocado/pull/4731#issuecomment-868683661)
> I have added a new commit with a fix for debian and ubuntu, I don't know how to fix it in windows.

I also see a failure on the rawhide RPM build:

```
check-884-selftests/functional/test_streams.py:StreamsTest.test_custom_stream_and_level FAIL
```

I'm investigating.

clebergnu commented:
> > > @clebergnu I changed `rglob` by `glob`, that should keep the current behavior in the Makefile
> > 
> > 
> > I've resubmitted a few failing CI jobs (failed when communicating with PyPI).
> 
> You have an explanation to that problem in a previous comment [#4731 (comment)](https://github.com/avocado-framework/avocado/pull/4731#issuecomment-868683661)
> I have added a new commit with a fix for debian and ubuntu, I don't know how to fix it in windows.

For Windows, it looks like the fix *may be* in newer versions of pip itself.  Can you please try to bump the Python installation to 3.9.5 on Windows (https://community.chocolatey.org/packages/python3/3.9.5), that is:

```diff
diff --git a/.cirrus.yml b/.cirrus.yml
index 38b088afd..dc964a4c3 100644
--- a/.cirrus.yml
+++ b/.cirrus.yml
@@ -55,7 +55,7 @@ debian_egg_task:
 
 windows_smokecheck_task:
     version_script:
-       - choco install -y python3 --version 3.9.0
+       - choco install -y python3 --version 3.9.5
        - C:\Python39\python setup.py develop --user
        - C:\Python39\python -m avocado --help
        - C:\Python39\python -m avocado run --test-runner=nrunner examples\tests\passtest.py

```

ana commented:
> > > > @clebergnu I changed `rglob` by `glob`, that should keep the current behavior in the Makefile
> > > 
> > > 
> > > I've resubmitted a few failing CI jobs (failed when communicating with PyPI).
> > 
> > 
> > You have an explanation to that problem in a previous comment [#4731 (comment)](https://github.com/avocado-framework/avocado/pull/4731#issuecomment-868683661)
> > I have added a new commit with a fix for debian and ubuntu, I don't know how to fix it in windows.
> 
> For Windows, it looks like the fix _may be_ in newer versions of pip itself. Can you please try to bump the Python installation to 3.9.5 on Windows (https://community.chocolatey.org/packages/python3/3.9.5), that is:
> 
> ```diff
> diff --git a/.cirrus.yml b/.cirrus.yml
> index 38b088afd..dc964a4c3 100644
> --- a/.cirrus.yml
> +++ b/.cirrus.yml
> @@ -55,7 +55,7 @@ debian_egg_task:
>  
>  windows_smokecheck_task:
>      version_script:
> -       - choco install -y python3 --version 3.9.0
> +       - choco install -y python3 --version 3.9.5
>         - C:\Python39\python setup.py develop --user
>         - C:\Python39\python -m avocado --help
>         - C:\Python39\python -m avocado run --test-runner=nrunner examples\tests\passtest.py
> ```

Thanks for the hint @clebergnu sadly, it didn't work.

ana commented:
@clebergnu Unless we get a different windows container image with updatesdcerts, I don't know how to solve the issue with windows and cirrus.  I've checked Chocolatey and there is nothing to update the certificates.
We're already checking windows in GHA, so an option is to remove windows from cirrus and leave only Fedora and Debian.

clebergnu commented:
> @clebergnu Unless we get a different windows container image with updatesdcerts, I don't know how to solve the issue with windows and cirrus. I've checked Chocolatey and there is nothing to update the certificates.
> We're already checking windows in GHA, so an option is to remove windows from cirrus and leave only Fedora and Debian.

Yes, they're overlapping.  Could you please remove the Windows job from Cirrus so that we have a passing CI on this PR?

Thanks!

ana commented:
> > @clebergnu Unless we get a different windows container image with updatesdcerts, I don't know how to solve the issue with windows and cirrus. I've checked Chocolatey and there is nothing to update the certificates.
> > We're already checking windows in GHA, so an option is to remove windows from cirrus and leave only Fedora and Debian.
> 
> Yes, they're overlapping. Could you please remove the Windows job from Cirrus so that we have a passing CI on this PR?
> 

done!

clebergnu commented:
> > > @clebergnu Unless we get a different windows container image with updatesdcerts, I don't know how to solve the issue with windows and cirrus. I've checked Chocolatey and there is nothing to update the certificates.
> > > We're already checking windows in GHA, so an option is to remove windows from cirrus and leave only Fedora and Debian.
> > 
> > 
> > Yes, they're overlapping. Could you please remove the Windows job from Cirrus so that we have a passing CI on this PR?
> 
> done!

Thanks!

clebergnu reviewed (approved):
LGTM.

PS: The failure on Travis/arm64 is not caused by this PR, appeared initially on 7774fac86b175d608a71bb03836e8e70e7b10e3b, but may also not be related to it.
